projects/sleuthkit/sleuthkit_mem_img.h

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MEM_IMG_H
+#define MEM_IMG_H
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sleuthkit/tsk/tsk_tools_i.h"
+
+typedef struct {
+  TSK_IMG_INFO img_info;
+  const uint8_t *data;
+  size_t size;
+} IMG_MEM_INFO;
+
+static ssize_t mem_read(TSK_IMG_INFO *img_info, TSK_OFF_T offset, char *buf,
+                        size_t len) {
+  IMG_MEM_INFO *mem_info =
+      reinterpret_cast<IMG_MEM_INFO*>(img_info);
+  // Bounds-checking exists in the real drivers.
+  if (offset > mem_info->size) {
+    return -1;
+  }
+  ssize_t read_len = len;
+  if (offset + len > mem_info->size) {
+    read_len = mem_info->size - offset;
+  }
+  if (memcpy(buf, mem_info->data + offset, read_len) == nullptr) {
+    return -1;
+  } else {
+    return read_len;
+  }
+}
+
+static void mem_close(TSK_IMG_INFO *img_info) {
+  IMG_MEM_INFO *mem_info = reinterpret_cast<IMG_MEM_INFO*>(img_info);
+  tsk_deinit_lock(&(img_info->cache_lock));
+  free(mem_info);
+}
+
+static void mem_imgstat(TSK_IMG_INFO *img_info, FILE *hFile) {}
+
+TSK_IMG_INFO *mem_open(const uint8_t *data, size_t size) {
+  IMG_MEM_INFO* inmemory_img;
+  TSK_IMG_INFO* img;
+  if ((inmemory_img = reinterpret_cast<IMG_MEM_INFO*>(
+       malloc(sizeof(IMG_MEM_INFO)))) == nullptr) {
```

Comment:
is it necessarily to use `malloc`? Maybe you can declare it on stack? That way fuzzing will be faster and the code will be simpler.

Reply:
I'm not sure this is possible. This header file has been taken from a set of [existing utilities,](https://github.com/google/security-research-pocs/blob/master/autofuzz/sleuthkit_mem_img.cc) and the `img` pointer that's returned is allocated in this function. The memory block here is being used in various scopes and across different files, so we should probably keep this instance of `malloc`.